spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,308 @@
+        RichIn(index, v)
+      case And(l, r) => And(richFilter(l), richFilter(r))
+      case Or(l, r) => Or(richFilter(l), richFilter(r))
+      case Not(f) => Not(richFilter(f))
+      case _ => throw new IllegalArgumentException(s"Unknown filter $filter")
+    }
+
+    /** Evaluate rich filter on a column value from a row */
+    def translateFilter(filter: Filter): Row => Boolean = filter match {
+      case RichEqualTo(colIndex, colDataType,  value) =>
+        (row: Row) => equalTo(row.get(colIndex), colDataType, value)
+      case RichLessThan(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) < 0
+      case RichLessThanOrEqual(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) <= 0
+      case RichGreaterThan(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) > 0
+      case RichGreaterThanOrEqual(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) >= 0
+      case RichIn(colIndex, values) => (row: Row) => values.toSet.contains(row.get(colIndex))
+      case RichIsNull(colIndex) => (row: Row) => row.get(colIndex) == null
+      case RichIsNotNull(colIndex) => (row: Row) => row.get(colIndex) != null
+      case Not(f) => (row: Row) => !translateFilter(f)(row)
+      case And(l, r) => (row: Row) => translateFilter(l)(row) && translateFilter(r)(row)
+      case Or(l, r) => (row: Row) => translateFilter(l)(row) || translateFilter(r)(row)
+      case _ => (row: Row) => throw new RuntimeException(s"Unknown RichFilter $filter")
+    }
+
+    // a filter combining all other filters
+    val richFilters = preservedFilters.map(richFilter)
+    val translatedFilters = richFilters.map(translateFilter)
+    def rowFilter(row: Row): Boolean = translatedFilters.forall(_(row))
+
+    val transformer = new RddTransformer(requiredColumns, pushdownFilters)
+    transformer.transform(baseRdd).filter(rowFilter)
+  }
+}
+
+object CassandraSourceRelation {
+
+  /** If push down is disable, use [[PrunedScan]]. By default use [[PrunedFilteredScan]]*/
+  def apply(tableIdent: TableIdent, sqlContext: SQLContext)(
+    implicit
+      connector: CassandraConnector =
+      new CassandraConnector(sqlContext.getCassandraConnConf(sqlContext.getCluster)),
+      readConf: ReadConf = sqlContext.getReadConf(tableIdent),
+      writeConf: WriteConf = sqlContext.getWriteConf(tableIdent),
+      sourceOptions: CassandraDataSourceOptions = CassandraDataSourceOptions()) : BaseRelationImpl = {
+
+    // Default scan type
+    if (sourceOptions.pushdown) {
+      new PrunedFilteredScanRelationImpl(
+        tableIdent = tableIdent,
+        connector = connector,
+        readConf = readConf,
+        writeConf = writeConf,
+        userSpecifiedSchema = sourceOptions.schema,
+        sqlContext = sqlContext)
+    } else {
+      new PrunedScanRelationImpl(
```

Comment:
Both `PrunedScanRelationImpl` and `PrunedFilteredScanRelationImpl` are very similar in implementation; the only difference is that one of them is doing predicate pushdown, while another one is not. So why not merge them into a single class with optional predicate pushdown enabled by a parameter?